tokio-trace/src/lib.rs

```diff
@@ -0,0 +1,774 @@
+//!
+//! ### In libraries
+//!
+//! Libraries should link only to the `tokio-trace` crate, and use the provided
+//! macros to record whatever information will be useful to downstream consumers.
+//!
+//! ### In executables
+//!
+//! In order to record trace events, executables have to use a `Subscriber`
+//! implementation compatible with `tokio-trace`. A `Subscriber` implements a way of
+//! collecting trace data, such as by logging it to standard output.
+//!
+//! Unlike the `log` crate, `tokio-trace` does *not* use a global `Subscriber` which
+//! is initialized once. Instead, it follows the `tokio` pattern of executing code
+//! in a context. For example:
+//!
+//! ```rust
+//! #[macro_use]
+//! extern crate tokio_trace;
+//! # pub struct FooSubscriber;
+//! # use tokio_trace::{span::Id, field::Field, Metadata};
+//! # impl tokio_trace::Subscriber for FooSubscriber {
+//! #   fn new_span(&self, _: &Metadata) -> Id { Id::from_u64(0) }
+//! #   fn record_debug(&self, _: &Id, _: &Field, _: &::std::fmt::Debug) {}
+//! #   fn add_follows_from(&self, _: &Id, _: Id) {}
+//! #   fn enabled(&self, _: &Metadata) -> bool { false }
+//! #   fn enter(&self, _: &Id) {}
+//! #   fn exit(&self, _: &Id) {}
+//! # }
+//! # impl FooSubscriber {
+//! #   fn new() -> Self { FooSubscriber }
+//! # }
+//! # fn main() {
+//! let my_subscriber = FooSubscriber::new();
+//!
+//! tokio_trace::subscriber::with_default(my_subscriber, || {
+//!     // Any trace events generated in this closure or by functions it calls
+//!     // will be collected by `my_subscriber`.
+//! })
+//! # }
+//! ```
+//!
+//! This approach allows trace data to be collected by multiple subscribers within
+//! different contexts in the program. Alternatively, a single subscriber may be
+//! constructed by the `main` function and all subsequent code executed with that
+//! subscriber as the default. Any trace events generated outside the context of a
+//! subscriber will not be collected.
+//!
+//! The executable itself may use the `tokio-trace` crate to instrument itself as well.
+//!
+//! [`log`]: https://docs.rs/log/0.4.6/log/
+//! [`Span`]: span/struct.Span
+//! [`Event`]: struct.Event.html
+//! [`Subscriber`]: subscriber/trait.Subscriber.html
+//! [`observe_event`]: subscriber/trait.Subscriber.html#tymethod.observe_event
+//! [`enter`]: subscriber/trait.Subscriber.html#tymethod.enter
+//! [`exit`]: subscriber/trait.Subscriber.html#tymethod.exit
+//! [`enabled`]: subscriber/trait.Subscriber.html#tymethod.enabled
+//! [metadata]: struct.Metadata.html
+extern crate tokio_trace_core;
+
+// Somehow this `use` statement is necessary for us to re-export the `core`
+// macros on Rust 1.26.0. I'm not sure how this makes it work, but it does.
+#[allow(unused_imports)]
+use tokio_trace_core::*;
+
+pub use self::{
+    dispatcher::Dispatch,
+    field::Value,
+    span::{Event, Id, Span},
+    subscriber::Subscriber,
+    tokio_trace_core::{
```

Comment:
Most of these probably should be `doc(hidden)`

Reply:
Done in 9478dbe1cc337b858b68814754206fc7d3393fd9.